rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/TextDocumentState.java

```diff
@@ -45,64 +59,253 @@
+                        }
+                    });
+            }
+        }
+    }
+}
+
+/**
+ * A *debouncer* is an object to get a *resource* from an *underlying resource
+ * provider* with a certain delay. From the perspective of the debouncer, the
+ * underlying resource provider has two states: initialized and not-initialized.
+ *
+ *  1. While the underlying resource provider is not-initialized (e.g., the
+ *     computation of a parse tree has not yet started), the debouncer waits
+ *     until the delay is over.
+ *
+ *  2. When the underlying resource provider becomes initialized (e.g., the
+ *     computation of a parse tree has started, but possibly not yet finished),
+ *     the debouncer returns a future for the resource.
+ *
+ *  3. When the underlying resource provider is not-initialized, but the delay
+ *     is over, the debouncer forcibly initializes the resource (e.g., it starts
+ *     the asynchronous computation of a parse tree) and returns a future for
+ *     the resource.
+ */
+class Debouncer<T> {
+
+    // A debouncer is implemented using a *delayed executor* as `scheduler`. The
+    // idea is to *periodically* check the state of the underlying resource
+    // provider. More precisely, each time when the resource is requested,
+    // immediately check if case 2 or case 3 (above) are applicable. If so,
+    // return. If not, schedule a *delayed future* to retry the request, to be
+    // completed after a small `period` (e.g., 50 milliseconds).
+    //
+    // The reason why multiple futures are scheduled in small periods, instead
+    // of a single future for the entire large delay, is that futures (of type
+    // `CompletableFuture`) cannot be interrupted.
+
+    private final int period; // Milliseconds
+    private final Executor scheduler;
+
+    // At any point in time, only one delayed future to retry the request for
+    // the resource should be `scheduled`, tied with the total remaining delay.
+    // For bookkeeping, a *stamped reference* is used. The reference is the
+    // delayed future, while the stamp is the remaining delay *upon completion
+    // of the delayed future*.
+
+    private final AtomicStampedReference<@Nullable CompletableFuture<T>> scheduled;
+
+    // The underlying resource provider is represented abstractly in terms of
+    // two suppliers, each of which corresponds with a state of the underlying
+    // resource provider. `getIfInitialized` should return `null` iff the
+    // underlying resource provider is not-initialized.
+
+    private final Supplier<@Nullable CompletableFuture<T>> getIfInitialized;
+    private final Supplier<CompletableFuture<T>> initializeAndGet;
+
+    public Debouncer(Duration period,
+            Supplier<@Nullable CompletableFuture<T>> getIfInitialized,
+            Supplier<CompletableFuture<T>> initializeAndGet) {
+
+        this(Math.toIntExact(period.toMillis()), getIfInitialized, initializeAndGet);
+    }
+
+    public Debouncer(int period,
```

Comment:
Can we write some junit tests for these classes? Like to make sure it behaves as intended, and later, people don't accidentally break it?